src/Controller/VerifyEmailControllerTrait.php

```diff
@@ -0,0 +1,46 @@
+<?php
+
+/*
+ * This file is part of the SymfonyCasts VerifyEmailBundle package.
+ * Copyright (c) SymfonyCasts <https://symfonycasts.com/>
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace SymfonyCasts\Bundle\VerifyEmail\Controller;
+
+use Symfony\Component\HttpFoundation\Request;
+
+/**
+ * Provides useful methods to a "verify email controller".
+ *
+ * Use of this trait requires a controller to extend
+ * Symfony\Bundle\FrameworkBundle\Controller\AbstractController
+ *
+ * @author Jesse Rushlow <jr@rushlow.dev>
+ * @author Ryan Weaver   <ryan@symfonycasts.com>
+ */
+class VerifyEmailControllerTrait
+{
+    private function isConfirmationValid(Request $request): void
+    {
+        $user = $this->getUser();
+
+        if ($user->isVerified()) {
+            throw new \Exception('Your email address has already been verified.');
+        }
```

Comment:
I don't think we care about this case. The user could choose to check this already if they wanted. But if the user is already verified... and they want to verify again, go for it! Technically speaking, this has nothing to do with the signature being valid or not.